src/Middleware/RequestBodyBufferMiddleware.php

```diff
@@ -30,27 +30,35 @@ public function __construct($sizeLimit = null)
 
     public function __invoke(ServerRequestInterface $request, $stack)
     {
+        $sizeLimit = $this->sizeLimit;
         $body = $request->getBody();
 
         // request body of known size exceeding limit
         if ($body->getSize() > $this->sizeLimit) {
-            return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
+            $sizeLimit = 0;
         }
 
         if (!$body instanceof ReadableStreamInterface) {
             return $stack($request);
         }
 
-        return Stream\buffer($body, $this->sizeLimit)->then(function ($buffer) use ($request, $stack) {
+        return Stream\buffer($body, $sizeLimit)->then(function ($buffer) use ($request, $stack) {
             $stream = new BufferStream(strlen($buffer));
             $stream->write($buffer);
             $request = $request->withBody($stream);
 
             return $stack($request);
-        }, function($error) {
+        }, function ($error) use ($stack, $request, $body) {
             // request body of unknown size exceeding limit during buffering
             if ($error instanceof OverflowException) {
-                return new Response(413, array('Content-Type' => 'text/plain'), 'Request body exceeds allowed limit');
+                return new Promise(function ($resolve, $reject) use ($stack, $request, $body) {
+                    $body->on('error', function ($error) use ($reject) {
+                        $reject($error);
+                    });
+                    $body->on('close', function () use ($stack, $request, $resolve) {
+                        $resolve($stack($request));
+                    });
+                });
```

Comment:
Changes LGTM, but this is currently missing a cancellation handler. Maybe use [`Stream\first()`](https://github.com/reactphp/promise-stream#first) instead?

Reply:
Good catch, I'll look into it :+1:

Reply:
Not sure how relevant this here, what do you think about https://github.com/reactphp/promise-stream/pull/7?